README.md
## Join-Object
Combines two object lists based on a related property between them.

## Description
Combines properties from one or more objects. It creates a set that can be saved as a new object or used as it is. An object join is a means for combining properties from one (self-join) or more object lists by using values common to each.

**Main features:**
* Intuitive (SQL like) syntax
* Smart property merging
* Predefined join commands for updating, merging and specific join types
* Well defined pipeline for the (left) input objects and output objects (preserves memory when correctly used)
* Performs about 40% faster than Compare-Object on large object lists
* Supports a list of (custom) objects, strings or primitives and dictionaries (e.g. hash tables) and data tables for input
* Smart properties and calculated property expressions
* Custom relation expressions
* Easy installation (dot-sourcing)
* Supports PowerShell for Windows (5.1) and PowerShell Core

The Join-Object cmdlet reveals the following proxy commands with their own (`-JoinType` and `-Property`) defaults:
* `InnerJoin-Object` (Alias `InnerJoin` or `Join`), combines the related objects
* `LeftJoin-Object` (Alias `LeftJoin`), combines the related objects and adds the rest of the left objects
* `RightJoin-Object` (Alias `RightJoin`), combines the related objects and adds the rest of the right objects
* `FullJoin-Object` (Alias `FullJoin`), combines the related objects and adds the rest of the left and right objects
* `CrossJoin-Object` (Alias `CrossJoin`), combines each left object with each right object
* `Update-Object` (Alias `Update`), updates the left object with the related right object
* `Merge-Object` (Alias `Merge`), updates the left object with the related right object and adds the rest of the
  new (unrelated) right objects

## Installation
There are two versions of this `Join-Object` cmdlet (both versions supply the same functionality):

* [**Join Module**](https://www.powershellgallery.com/packages/JoinModule)

<span></span>

    Install-Module -Name JoinModule

* [**Join Script**](https://www.powershellgallery.com/packages/Join)

<span></span>

    Install-Script -Name Join

(Or rename the `Join.psm1` module to a `Join.ps1` script file) and invoked the script by [**dot sourcing**](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_scripts?view=powershell-6#script-scope-and-dot-sourcing): 

    . .\Join.ps1`

## Examples

**Example 1**
```PowerShell
PS C:\> $Employee

Id Name    Country Department  Age ReportsTo
-- ----    ------- ----------  --- ---------
 1 Aerts   Belgium Sales        40         5
 2 Bauer   Germany Engineering  31         4
 3 Cook    England Sales        69         1
 4 Duval   France  Engineering  21         5
 5 Evans   England Marketing    35
 6 Fischer Germany Engineering  29         4

PS C:\> $Department

Name        Country
----        -------
Engineering Germany
Marketing   England
Sales       France
Purchase    France


PS C:\> # Join the employees with the departments based on the country
PS C:\> $Employee | InnerJoin $Department -On Country | Format-Table

Id Name                   Country Department  Age ReportsTo
-- ----                   ------- ----------  --- ---------
 2 {Bauer, Engineering}   Germany Engineering  31         4
 3 {Cook, Marketing}      England Sales        69         1
 4 {Duval, Sales}         France  Engineering  21         5
 4 {Duval, Purchase}      France  Engineering  21         5
 5 {Evans, Marketing}     England Marketing    35
 6 {Fischer, Engineering} Germany Engineering  29         4
```

**Example 2**
```PowerShell
PS C:\> # Full join the employees with the departments based on the department name
PS C:\> # and Split the names over differend properties
PS C:\> $Employee | InnerJoin $Department -On Department -Equals Name -Discern Employee, Department | Format-Table

Id Name    EmployeeCountry DepartmentCountry Department  Age ReportsTo
-- ----    --------------- ----------------- ----------  --- ---------
 1 Aerts   Belgium         France            Sales        40         5
 2 Bauer   Germany         Germany           Engineering  31         4
 3 Cook    England         France            Sales        69         1
 4 Duval   France          Germany           Engineering  21         5
 5 Evans   England         England           Marketing    35
 6 Fischer Germany         Germany           Engineering  29         4
```

**Example 3**
```PowerShell
PS C:\> $Changes

Id Name    Country Department  Age ReportsTo
-- ----    ------- ----------  --- ---------
 3 Cook    England Sales        69         5
 6 Fischer France  Engineering  29         4
 7 Geralds Belgium Sales        71         1


PS C:\> # Apply the changes to the employees
PS C:\> $Employee | Merge $Changes -On Id | Format-Table

Id Name    Country Department  Age ReportsTo
-- ----    ------- ----------  --- ---------
 1 Aerts   Belgium Sales        40         5
 2 Bauer   Germany Engineering  31         4
 3 Cook    England Sales        69         5
 4 Duval   France  Engineering  21         5
 5 Evans   England Marketing    35
 6 Fischer France  Engineering  29         4
 7 Geralds Belgium Sales        71         1
```

**Example 4**
```PowerShell
PS C:\> # (Self) join each employee with its each manager
PS C:\> LeftJoin $Employee -On ReportsTo -Equals Id -Property @{ Name = 'Left.Name' }, @{ Manager = 'Right.Name' }

Name    Manager
----    -------
Aerts   Evans
Bauer   Duval
Cook    Aerts
Duval   Evans
Evans
Fischer Duval
```

**Example 5**
```PowerShell
PS C:\> # Add an Id to the department list
PS C:\> 1..9 |Join $Department -ValueName Id

Id Name        Country
-- ----        -------
 1 Engineering Germany
 2 Marketing   England
 3 Sales       France
 4 Purchase    France
```

**Example 6**
```PowerShell
PS C:\> $a = 'a1', 'a2', 'a3', 'a4'
PS C:\> $b = 'b1', 'b2', 'b3', 'b4'
PS C:\> $c = 'c1', 'c2', 'c3', 'c4'
PS C:\> $d = 'd1', 'd2', 'd3', 'd4'

PS C:\> # Join (transpose) multiple arrays to a collection array
PS C:\> $a |Join $b |Join $c |Join $d |% { "$_" }

a1 b1 c1 d1
a2 b2 c2 d2
a3 b3 c3 d3
a4 b4 c4 d4
```

**Example 7**
```PowerShell
PS C:\> # Create objects with named properties from multiple arrays
PS C:\> $a |Join $b |Join $c |Join $d -Name a, b, c, d

a  b  c  d
-  -  -  -
a1 b1 c1 d1
a2 b2 c2 d2
a3 b3 c3 d3
a4 b4 c4 d4
```

## Parameters

**`-LeftObject <object list, data table or list of hash tables>`**  
The left object list, usually provided through the pipeline, to be joined.

*Note:* a self-join on the LeftObject list will be performed if the RightObject is omitted.

**`-RightObject <object list, data table or list of hash tables>`**  
The right object list, provided by the first argument, to be joined.

*Note:* a self-join on the RightObject list will be performed if the LeftObject is omitted.

**`-On <String[]>`**  
The `-On` parameter (alias `-Using`) defines which objects should be joined together.
If the `-Equals` parameter is omitted, the value(s) of the properties listed by the `-On` parameter should be equal at both sides in order to join the left object with the right object.

*Note 1:* The list of properties defined by the `-On` parameter will be complemented with the list of
properties defined by the `-Equals` parameter and vice versa.

*Note 2:* Related properties will be merged to a single property by default (see also the -Property
parameter).

*Note 3:* If the -On and the `-OnExpression` parameter are omitted, a side-by-side join is returned.

**`-Equals <String[]>`**  
If the `-Equals` parameter is supplied, the value(s) of the left object properties listed by the `-On`
parameter should be equal to the value(s)of the right object listed by the `-Equals` parameter in order to join the left object with the right object.

*Note 1:* The list of properties defined by the `-Equal` parameter will be complemented with the list of properties defined by the `-On` parameter and vice versa.

*Note 2:* A property will be omitted if it exists on both sides and if the property at the other side is
related to another property.

*Note 3:* The `-Equals` parameter can only be used with the `-On` parameter.

**`-Strict`**  
If the `-Strict` switch is set, the comparison between the related properties defined by the `-On` Parameter (and the `-Equals` parameter) is based on a strict equality (both type and value need to be equal).

**`-MatchCase`**  
If the `-MatchCase` (alias `-CaseSensitive`) switch is set, the comparison between the related properties defined by the `-On` Parameter (and the `-Equals` parameter) will case sensitive.

**`-OnExpression <ScriptBlock>`**  
Any conditional expression (where `$Left` refers to each left object and `$Right` refers to each right object) that requires to evaluate to true in order to join the left object with the right object.

*Note 1:* The `-OnExpression` parameter has the most complex comparison possibilities but is considerable slower than the other types.

*Note 2:* The `-OnExpression` parameter cannot be used with the `-On` parameter.

**`-Where <ScriptBlock>`**  
An expression that defines the condition to be met for the objects to be returned. There is no limit to the number of predicates that can be included in the condition.

**`-Discern <String[]>`**  
By default unrelated properties with the same name will be collected in a single object property.
The `-Discern` parameter (alias `-NameItems`)  defines how to rename the object properties and divide them over multiple properties. If a given name pattern contains an asterisks (`*`), the asterisks will be replaced with the original property name. Otherwise, the property name for each property item will be prefixed with the given name pattern.

The property collection of multiple (chained) join commands can be divided in once from the last join command in the change. The rename patterns are right aligned, meaning that the last renamed pattern will be applied to the last object joined. If there are less rename patterns than property items, the rest of the (left most) property items will be put in a fixed array under the original property name.

*Note 1:* Only properties with the same name on both sides will not be renamed.

*Note 2:* Related properties (with an equal value defined by the -On parameter) will be merged to a signle
item.

**`-Property <(HashTable or String)[]>`**  
A hash table or list of property names (strings) and/or hash tables that define a new selection of
property names and values

Hash tables should be in the format `@{<PropertyName> = <Expression>}` where the <Expression> is a `ScriptBlock` or a *smart property* (string) and defines how the specific left and right properties should be merged.

The following variables are exposed for a (`ScriptBlock`) expression:
* **`$_`**: iterates each property name
* **`$Left`**: a hash table representing the current left object (each self-contained `-LeftObject`).
  The hash table will be empty (`@{}`) in the outer part of a left join or full join.
* **`$LeftIndex`**: the index of the left object (`$Null` in the outer part of a right- or full join)
* **`$Right`**: a hash table representing the current right object (each self-contained `-RightObject`)
  The hash table will be empty (`@{}`) in the outer part of a right join or full join.
* **`$RightIndex`**: the index of the right object (`$Null` in the outer part of a left- or full join)

The following smart properties are available:
* A **general property**: `'<Property Name>'`, where `<Property Name>` represents the property name of the left and/or right property, e.g. `@{ MyProperty = 'Name' }`. If the property exists on both sides, an array holding both values will be returned. In the outer join, the value of the property will be `$Null`.  This smart property is similar to the expression: `@{ MyProperty = { @($Left['Name'], $Right['Name']) } }`
* A **general wildcard property**: `'*'`, where `*` represents the property name of the current property, e.g. ` 'MyProperty' in @{ MyProperty = '*' }`. If the property exists on both sides:
  - and the properties are unrelated, an array holding both values will be returned
  - and the properties are related to each other, the (equal) values will be merged in one property value
  - and the property at the other side is related to an different property, the property is omitted
  
  The argument: `-Property *`, will apply a general wildcard on all left and right properties.
* A **left property**: `'Left.<Property Name>'`, or **right property**: `'Right.<Property Name>'`, where `<Property Name>` represents the property name of the either the left or right property. If the property doesn't exist, the value of the property will be `$Null`.
* A **left wildcard property**: `'Left.*'`, or **right wildcard property**: `'Right.*'`, where `*` represents the property name of the current the left or right property, e.g. `'MyProperty' in @{ MyProperty = 'Left.*' }`. If the property doesn't exist (in an outer join), the property with the same name at the other side will be taken. If the property doesn't exist on either side, the value of the property will be `$Null`.
  
  The argument: `-Property 'Left.*'`, will apply a left wildcard property on all the left object properties.

If the `-Property` parameter and the `-Discern` parameter are omitted, a general wildcard property is applied on all the left and right properties.

The last defined expression or smart property will overrule any previous defined properties.

**`-ValueName <String>`**  
Defines the default name for the property name in case a scalar array is joined with an object array.

**`-JoinType <'Inner'|'Left'|'Right'|'Full'|'Cross'>`**  
Defines which unrelated objects should be included (see: Description). The default is `'Inner'`.

Note: It is recommended to use the related proxy commands (`... | <JoinType>-Object ...`) instead.

<sub>Please give a 👍 if you support the proposal to [Add a Join-Object cmdlet to the standard PowerShell equipment (`#14994`)](https://github.com/PowerShell/PowerShell/issues/14994)</sub>


CHANGELOG
[Version] [DateTime] Author      Comments
--------- ---------- ------      --------
00.99.99  2017-01-01 Ronald Bode First releases 
01.01.00  2017-08-08 Ronald Bode Merged the -Expressions and -DefaultExpression parameters
01.01.01  2017-10-24 Ronald Bode Resolved bug where the Left Table contains a single column
02.00.02  2017-12-11 Ronald Bode Reworked for PowerSnippets.com
02.01.00  2018-03-01 Ronald Bode Resolved: "Unexpected results when reusing custom objects in the pipeline"
02.02.00  2018-03-15 Ronald Bode Read single records from the pipeline
02.02.01  2018-03-25 Ronald Bode Support for adding new properties (see: -Property)
2.2.2     2018-03-25 Ronald Bode PowerShell Gallery Release
2.2.3     2018-03-25 Ronald Bode Supply a list of properties by: -Property [String[]]
2.2.4     2018-03-25 Ronald Bode Keeping the properties in order
2.2.5     2018-11-27 Ronald Bode Fixed empty output bug (including test)
2.2.6     2018-11-28 Ronald Bode Support for mixed [string]Key/[hashtable]@{Key={Expression}} -Property parameter
2.3.0     2018-11-28 Ronald Bode Replaced InnerJoin-, LeftJoin-, RightJoin-, FullJoin-Object aliases by proxy commands
2.3.1     2018-12-30 Ronald Bode Added CrossJoin Type. If the -On parameter is omitted, a join by index will be done
2.3.2     2019-02-01 Ronald Bode The -MergeExpression is only used in case the Left and Right properties overlap
2.4.0     2019-02-02 Ronald Bode Added Update-Object and Merge-Object proxy commands
2.4.1     2019-02-06 Ronald Bode Added $LeftOrRight and $RightOrLeft references
2.4.2     2019-02-08 mcclimont   Compliant with StrictMode -Version 2 (https://github.com/iRon7/Join-Object/pull/3)
2.4.3     2019-02-06 Ronald Bode Changed $LeftOrNull and $RightOrNull to $LeftOrVoid and $RightOrVoid
2.4.4     2019-02-24 Ronald Bode Resolved scope bug when invoked multiple times in the same stream
2.5.1     2019-03-03 Ronald Bode Added -Pair (alias -Merge) feature to separate duplicated unrelated property names
2.5.2     2019-03-04 Ronald Bode Changed -Pair to -Unify
2.6.0     2019-03-10 Ronald Bode Improved performance by using a HashTable for the inner (right) loop where possible
3.7.0     2019-03-25 Ronald Bode Added -Where clause
3.7.1     2019-03-29 Ronald Bode Improved self join syntax
3.0.0     2019-03-30 Ronald Bode New release with new test set
3.0.1     2019-03-30 Ronald Bode Updated embedded examples
3.0.2     2019-04-10 Ronald Bode Fixed default unify issue due to -On case difference
3.0.3     2019-07-02 Ronald Bode Support for datatables
3.0.4     2019-07-03 Ronald Bode Experimental version (not implemented)
3.0.5     2019-07-14 meany       Issue #5, resolved: Cannot dot source / invoke script on 2012 R2 bug
3.0.6     2019-07-16 Ronald Bode Issue #6, improved performance (~2x on large tables), thanks to @burkasaurusrex' suggestion
3.0.7     2019-11-01 Ronald Bode Renamed -Unify parameter to -Discern and divided -Discern from -Property parameter
3.0.8     2019-11-07 Ronald Bode Better support chaining multiple joins and simplified available expression objects: 
                                 All properties of the $Left and $Right object are set to $Null in the outer join part.
3.1.0     2019-11-08 Ronald Bode Adjusted help
3.1.1     2019-11-10 Ronald Bode Improved -Property * implementation 
3.1.2     2019-11-11 Ronald Bode Resolved bug with single right object
3.1.3     2019-11-15 Ronald Bode Also apply -Where argument to outer join part (expression to evaluate $Null values)
3.1.4     2019-12-02 Ronald Bode Throw "The -On parameter cannot be used on a cross join."
3.1.5     2019-12-09 Ronald Bode Added -Strict parameter
3.1.6     2019-12-10 Ronald Bode Added -MatchCase (alias -CaseSensitive) parameter
3.2.0     2019-12-16 Ronald Bode Defined stricter parameter sets (separated -On <String[]> and -OnExpression <ScriptBlock>)
3.2.1     2020-01-19 Ronald Bode Issue #10: Support for dictionaries (hashtable, ordered, ...)
3.2.2     2020-04-05 Ronald Bode Better handling argument exceptions
3.3.0     2020-08-09 Ronald Bode Reformatted script with https://github.com/DTW-DanWard/PowerShell-Beautifier
                                 Implemented smarter properties merge: https://github.com/iRon7/Join-Object/issues/12
                                 Prevent against code injection: https://devblogs.microsoft.com/powershell/powershell-injection-hunter-security-auditing-for-powershell-scripts/
                                 Support wildcard * (all properties) for the -On parameter
                                 Convert each object to a hash table for strict expressions
3.4.0     2021-03-01 Ronald Bode Implemented issue #14 "Support non-object arrays"
3.4.1     2021-03-08 Ronald Bode Implemented issue #16 "Discern merged properties for multiple joins"
3.4.2     2021-03-11 Ronald Bode Code and Help clearance
3.4.2     2021-03-20 Ronald Bode Issue #18 Support self-join on the left (piped) object
3.4.4     2021-03-24 Ronald Bode Using $PSCmdlet.ThrowTerminatingError for argument exceptions
3.4.5     2021-03-11 Ronald Bode Issue #17 Updated help
3.4.6     2021-04-08 Ronald Bode Fixed issue #19: Deal with empty (and $Null) inputs
3.4.7     2021-04-08 Ronald Bode Improved proxy command defaults
3.5.0     2021-06-08 Ronald Bode Prepared for module version
3.5.1     2021-06-10 Ronald Bode Fixed ScriptBlock module scope issue: https://stackoverflow.com/q/2193410/1701026
3.5.2     2021-06-14 Ronald Bode Minor Help update and advertisement for the Module version.
3.5.3     2021-07-06 Ronald Bode Fixed issue #27: MissingLeftProperty: Join-Object : The property 'xxx' cannot be found on the left object.
3.5.4     2021-07-27 Ronald Bode Fixed issue #28: FullJoin doesn't work properly when joining multiple array when one of the array is empty

Tests/Performance.Tests.ps1
Write-Host 'Preparing...'
. .\Join-Object.ps1
$Null = Random -SetSeed 0
$Max = 1000
$NumberFormat = '{0:' + '0' * ($Max - 1).ToString().Length + '}'

$Left = foreach ($i in 0..$Max) {
    [pscustomobject]@{
        Number = $i
        Random = "Random$NumberFormat" -f (Random $Max)
        Side = "Left$NumberFormat" -f $i
    }
}

$Right = foreach ($i in 0..$Max) {
    [pscustomobject]@{
        Name = "Name$NumberFormat" -f $i
        Random = "Random$NumberFormat" -f (Random $Max)
        Side = "Right$NumberFormat" -f $i
    } 
}


Write-Host 'Measuring inner join...'
$InnerJoin = Measure-Command {
    $Test = $Left | Join $Right -On Random
}
Write-Host ($Test | Select-Object -First 5 | Format-Table | Out-String)
Write-Host $InnerJoin.TotalSeconds

Write-Host 'Measuring full join...'
$FullJoin = Measure-Command {
    $Test = $Left | FullJoin $Right -On Random
}
Write-Host ($Test | Select-Object -First 5 | Format-Table | Out-String)
Write-Host $FullJoin.TotalSeconds

Write-Host 'Measuring side-by-side join...'
$SideJoin = Measure-Command {
    $Test = $Left | FullJoin $Right
}
Write-Host ($Test | Select-Object -First 5 | Format-Table | Out-String)
Write-Host $SideJoin.TotalSeconds

Write-Host 'Total:' ($InnerJoin.TotalSeconds + $FullJoin.TotalSeconds + $SideJoin.TotalSeconds)


Join-Object.Tests.ps1
$here = Split-Path -Parent $MyInvocation.MyCommand.Path
$sut = (Split-Path -Leaf $MyInvocation.MyCommand.Path).Replace(".Tests.", ".")
. "$here\$sut"

. .\ConvertFrom-SourceTable.ps1			# https://www.powershellgallery.com/packages/ConvertFrom-SourceTable

Function Should-BeObject {
	Param (
		[Parameter(Position=0)][Object[]]$b, [Parameter(ValueFromPipeLine = $True)][Object[]]$a
	)
	$Property = ($a | Select-Object -First 1).PSObject.Properties | Select-Object -Expand Name
	$Difference = Compare-Object $b $a -Property $Property
	Try {"$($Difference | Select-Object -First 1)" | Should -BeNull} Catch {$PSCmdlet.WriteError($_)}

}

Describe 'Join-Object' {
	
	$Employee = ConvertFrom-SourceTable '
		Name    Country Department
		----    ------- ----------
		Aerts   Belgium Sales
		Bauer   Germany Engineering
		Cook    England Sales
		Duval   France  Engineering
		Evans   England Marketing
		Fischer Germany Engineering'

	$Department = ConvertFrom-SourceTable '
		Name        Country     Manager
		----        -------     -------
		Engineering Germany     Meyer
		Marketing   England     Morris
		Sales       France      Millet
		Board       Netherlands Mans'
	
	Context 'Join types' {

		It 'InnerJoin' {
			,($Employee | InnerJoin $Department -On Country) | Should-BeObject (ConvertFrom-SourceTable '
				Country Department  Manager                     Name
				------- ----------  -------                     ----
				Germany Engineering Meyer     "Bauer", "Engineering"
				England Sales       Morris       "Cook", "Marketing"
				France  Engineering Millet          "Duval", "Sales"
				England Marketing   Morris      "Evans", "Marketing"
				Germany Engineering Meyer   "Fischer", "Engineering"')
		}

		It 'LeftJoin' {
			,($Employee | LeftJoin $Department -On Country) | Should-BeObject (ConvertFrom-SourceTable '
				Country Department  Manager                     Name
				------- ----------  -------                     ----
				Belgium Sales         $Null           "Aerts", $Null
				Germany Engineering Meyer     "Bauer", "Engineering"
				England Sales       Morris       "Cook", "Marketing"
				France  Engineering Millet          "Duval", "Sales"
				England Marketing   Morris      "Evans", "Marketing"
				Germany Engineering Meyer   "Fischer", "Engineering"')
		}

		It "RightJoin" {
			,($Employee | RightJoin $Department -On Country) | Should-BeObject (ConvertFrom-SourceTable '
				Department                      Name Manager Country
				----------                      ---- ------- -------
				Engineering   "Bauer", "Engineering" Meyer   Germany
				Sales            "Cook", "Marketing" Morris  England
				Engineering         "Duval", "Sales" Millet  France
				Marketing       "Evans", "Marketing" Morris  England
				Engineering "Fischer", "Engineering" Meyer   Germany
				      $Null           $Null, "Board" Mans    Netherlands')
		}

		It 'FullJoin' {
			,($Employee | FullJoin $Department -On Country) | Should-BeObject (ConvertFrom-SourceTable '
				Country     Department                      Name Manager
				-------     ----------                      ---- -------
				Belgium     Sales                 "Aerts", $Null   $Null
				Germany     Engineering   "Bauer", "Engineering" Meyer
				England     Sales            "Cook", "Marketing" Morris
				France      Engineering         "Duval", "Sales" Millet
				England     Marketing       "Evans", "Marketing" Morris
				Germany     Engineering "Fischer", "Engineering" Meyer
				Netherlands       $Null        @($Null, "Board") Mans')
		}

		It 'Cross Join' {
			,($Employee | Join $Department) | Should-BeObject (ConvertFrom-SourceTable '
				                 Country Department  Manager                     Name
				                 ------- ----------  -------                     ----
				    "Belgium", "Germany" Sales       Meyer     "Aerts", "Engineering"
				    "Belgium", "England" Sales       Morris      "Aerts", "Marketing"
				     "Belgium", "France" Sales       Millet          "Aerts", "Sales"
				"Belgium", "Netherlands" Sales       Mans            "Aerts", "Board"
				    "Germany", "Germany" Engineering Meyer     "Bauer", "Engineering"
				    "Germany", "England" Engineering Morris      "Bauer", "Marketing"
				     "Germany", "France" Engineering Millet          "Bauer", "Sales"
				"Germany", "Netherlands" Engineering Mans            "Bauer", "Board"
				    "England", "Germany" Sales       Meyer      "Cook", "Engineering"
				    "England", "England" Sales       Morris       "Cook", "Marketing"
				     "England", "France" Sales       Millet           "Cook", "Sales"
				"England", "Netherlands" Sales       Mans             "Cook", "Board"
				     "France", "Germany" Engineering Meyer     "Duval", "Engineering"
				     "France", "England" Engineering Morris      "Duval", "Marketing"
				      "France", "France" Engineering Millet          "Duval", "Sales"
				 "France", "Netherlands" Engineering Mans            "Duval", "Board"
				    "England", "Germany" Marketing   Meyer     "Evans", "Engineering"
				    "England", "England" Marketing   Morris      "Evans", "Marketing"
				     "England", "France" Marketing   Millet          "Evans", "Sales"
				"England", "Netherlands" Marketing   Mans            "Evans", "Board"
				    "Germany", "Germany" Engineering Meyer   "Fischer", "Engineering"
				    "Germany", "England" Engineering Morris    "Fischer", "Marketing"
				     "Germany", "France" Engineering Millet        "Fischer", "Sales"
				"Germany", "Netherlands" Engineering Mans          "Fischer", "Board"')
		}

	}
	
	Context 'Single object' {

		It 'Single left object' {
			,($Employee[1] | InnerJoin $Department -On Country) | Should-BeObject (ConvertFrom-SourceTable '
				Country Department  Manager                   Name
				------- ----------  ------- ----------------------
				Germany Engineering Meyer   "Bauer", "Engineering"')
		}

		It 'Single right object' {
			,($Employee | InnerJoin $Department[0] -On Country) | Should-BeObject (ConvertFrom-SourceTable '
				Country Department  Manager                     Name
				------- ----------  ------- ------------------------
				Germany Engineering Meyer     "Bauer", "Engineering"
				Germany Engineering Meyer   "Fischer", "Engineering"')
		}

		It 'Single left object and single right object' {
			,($Employee[1] | InnerJoin $Department[0] -On Country) | Should-BeObject (ConvertFrom-SourceTable '
				Country Department  Manager                   Name
				------- ----------  ------- ----------------------
				Germany Engineering Meyer   "Bauer", "Engineering"')
		}
	}

	Context "-On ... -Equals ..." {

		It '$Employee | InnerJoin $Department -On Department -Eq Name' {
			,($Employee | InnerJoin $Department -On Department -Eq Name) | Should-BeObject (ConvertFrom-SourceTable '
				             Country Department                      Name Manager
				             ------- ----------- ------------------------ -------
				 "Belgium", "France" Sales               "Aerts", "Sales" Millet
				"Germany", "Germany" Engineering   "Bauer", "Engineering" Meyer
				 "England", "France" Sales                "Cook", "Sales" Millet
				 "France", "Germany" Engineering   "Duval", "Engineering" Meyer
				"England", "England" Marketing       "Evans", "Marketing" Morris
				"Germany", "Germany" Engineering "Fischer", "Engineering" Meyer')
		}
	}

	Context "Merge values on Department = Name" {

		It 'Use the left object property if exists otherwise use right object property' {
			,($Employee | InnerJoin $Department -On Department -Eq Name {If ($Null -ne $Left.$_) {$Left.$_} Else {$Right.$_}}) | Should-BeObject (ConvertFrom-SourceTable '
				Department  Name    Manager Country
				----------  ----    ------- -------
				Sales       Aerts   Millet  Belgium
				Engineering Bauer   Meyer   Germany
				Sales       Cook    Millet  England
				Engineering Duval   Meyer   France
				Marketing   Evans   Morris  England
				Engineering Fischer Meyer   Germany')
		}

		It 'Only use the left name property and the right manager property' {
			,($Employee | InnerJoin $Department -On Department -Eq Name -Property @{Name = {$Left.$_}; Manager = {$Right.$_}}) | Should-BeObject (ConvertFrom-SourceTable '
				Name    Manager
				----    -------
				Aerts   Millet
				Bauer   Meyer
				Cook    Millet
				Duval   Meyer
				Evans   Morris
				Fischer Meyer')
		}

		It 'Use the left object property except for the country property' {
			,($Employee | InnerJoin $Department -On Department -Eq Name {$Left.$_} @{Manager = {$Right.$_}}) | Should-BeObject (ConvertFrom-SourceTable '
				Department  Name    Manager Country
				----------  ----    ------- -------
				Sales       Aerts   Millet  Belgium
				Engineering Bauer   Meyer   Germany
				Sales       Cook    Millet  England
				Engineering Duval   Meyer   France
				Marketing   Evans   Morris  England
				Engineering Fischer Meyer   Germany')
		}
	}

	Context "Join using expression" {

		It 'InnerJoin on Employee.Department = Department.Name and Employee.Country = Department.Country' {
			,($Employee | InnerJoin $Department -Using {$Left.Department -eq $Right.Name -and $Left.Country -eq $Right.Country} {$Left.$_} @{Manager = {$Right.$_}}) | Should-BeObject (ConvertFrom-SourceTable '
				Department  Name    Manager Country
				----------  ----    ------- -------
				Engineering Bauer   Meyer   Germany
				Marketing   Evans   Morris  England
				Engineering Fischer Meyer   Germany')
		}

		It 'Inner join on index' {
			,($Employee | InnerJoin $Department {$LeftIndex -eq $RightIndex}) | Should-BeObject (ConvertFrom-SourceTable '
				                Country Department                    Name Manager
				 ---------------------- ----------- ---------------------- -------
				   "Belgium", "Germany" Sales       "Aerts", "Engineering" Meyer
				   "Germany", "England" Engineering   "Bauer", "Marketing" Morris
				    "England", "France" Sales              "Cook", "Sales" Millet
				"France", "Netherlands" Engineering       "Duval", "Board" Mans')
		}
		
		It 'Full join on index' {
			,($Employee | FullJoin $Department {$LeftIndex -eq $RightIndex}) | Should-BeObject (ConvertFrom-SourceTable '
				                Country Department                    Name Manager
				----------------------- ----------  ---------------------- -------
				   "Belgium", "Germany" Sales       "Aerts", "Engineering" Meyer
				   "Germany", "England" Engineering   "Bauer", "Marketing" Morris
				    "England", "France" Sales              "Cook", "Sales" Millet
				"France", "Netherlands" Engineering       "Duval", "Board" Mans
				    `  "England", $Null Marketing   `       "Evans", $Null   $Null
				`      "Germany", $Null Engineering `     "Fischer", $Null   $Null')
		}
		
		It 'Self join with new properties' {
			$Employees = ConvertFrom-SourceTable '
				EmployeeId LastName  FirstName ReportsTo
				---------- --------  --------- ---------
				         1 Davolio   Nancy              2
				         2 Fuller    Andrew
				         3 Leveling  Janet              2
				         4 Peacock   Margaret           2
				         5 Buchanan  Steven             2
				         6 Suyama    Michael            5
				         7 King      Robert             5
				         8 Callahan  Laura              2
				         9 Dodsworth Anne               5'
				
			,($Employees | InnerJoin $Employees -On ReportsTo -Eq EmployeeID -Property @{Name = {"$($Left.FirstName) $($Left.LastName)"}; Manager = {"$($Right.FirstName) $($Right.LastName)"}}) | Should-BeObject (ConvertFrom-SourceTable '
				Name             Manager
				----             -------
				Nancy Davolio    Andrew Fuller
				Janet Leveling   Andrew Fuller
				Margaret Peacock Andrew Fuller
				Steven Buchanan  Andrew Fuller
				Michael Suyama   Steven Buchanan
				Robert King      Steven Buchanan
				Laura Callahan   Andrew Fuller
				Anne Dodsworth   Steven Buchanan')
		}

		It 'InnerJoin using multiple property matches and output specific columns' {
			$Left = ConvertFrom-SourceTable '
				   Ref_ID First_Name Last_Name [DateTime]DOB
				--------- ---------- --------- -------------
				321364060 User1      Micah     11/01/1969
				946497594 User2      Acker     05/28/1960
				887327716 User3      Aco       06/26/1950
				588496260 User4      John      05/23/1960
				565465465 User5      Jack      07/08/2020'
				
			$Right = ConvertFrom-SourceTable '
				First_Name Last_Name [DateTime]DOB City   Document_Type Filename
				---------- --------- ------------- ------ ------------- ------------
				User1      Micah     11/01/1969    Parker Transcript    T4IJZSYO.pdf
				User2      Acker     05/28/1960           Transcript    R4IKTRYN.pdf
				User3      Aco       06/26/1950           Transcript    R4IKTHMK.pdf
				User4      John      05/23/1960           Letter        R4IKTHSL.pdf'
				
			$Expected = ConvertFrom-SourceTable '
				Filename     Document_Type    Ref_ID First_Name [DateTime]DOB Last_Name
				--------     -------------    ------ ---------- ------------- ---------
				T4IJZSYO.pdf Transcript    321364060 User1      1969/11/01    Micah
				R4IKTRYN.pdf Transcript    946497594 User2      1960/05/28    Acker
				R4IKTHMK.pdf Transcript    887327716 User3      1950/06/26    Aco
				R4IKTHSL.pdf Letter        588496260 User4      1960/05/23    John'
			
			$Actual = $Left | Join $Right -On First_Name, Last_Name, DOB -Property Ref_ID, Filename, Document_Type, First_Name, DOB, Last_Name

			,$Actual | Should-BeObject $Expected
		}
	}

}
